core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -133,28 +140,43 @@ public ByteBuf readBytes(int length) {
 
     private ByteBuf readBytesSlow(int length) {
         ByteBuf value = null;
-        int remained = length;
-        for (ByteBuf buf : queue) {
+        int remaining = length;
+        for (final Iterator<ByteBuf> it = queue.iterator(); it.hasNext();) {
+            final ByteBuf buf = it.next();
             if (!buf.isReadable()) {
+                it.remove();
+                buf.release();
                 continue;
             }
```

Comment:
We don't need this at all if we never add an empty buffer and always remove an empty buffer as soon as it's consumed. Just an assertion will do, e.g. `assert readableBytes > 0 : buf`